pkg/trace/api/normalizer.go

```diff
@@ -21,48 +23,78 @@ const (
 	MaxNameLen = 100
 	// MaxTypeLen the maximum length a span type can have
 	MaxTypeLen = 100
-	// MaxEndDateOffset the maximum amount of time in the future we
-	// tolerate for span end dates
-	MaxEndDateOffset = 10 * time.Minute
+	// DefaultServiceName is the default name we assign a service if it's missing and we have no reasonable fallback
+	DefaultServiceName = "service"
+	// DefaultSpanName is the default name we assign a span if it's missing and we have no reasonable fallback
+	DefaultSpanName = "service.trace"
 )
 
 var (
 	// Year2000NanosecTS is an arbitrary cutoff to spot weird-looking values
 	Year2000NanosecTS = time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC).UnixNano()
 )
 
-// normalize makes sure a Span is properly initialized and encloses the minimum required info
-func normalize(s *pb.Span) error {
+// normalize makes sure a Span is properly initialized and encloses the minimum required info, returning error if it
+// is invalid beyond repair
+func normalize(ts *info.TagStats, s *pb.Span) error {
+	fallbackServiceName := DefaultServiceName
+	if ts.Lang != "" {
+		fallbackServiceName = ts.Lang
+	}
+
+	if s.TraceID == 0 {
+		atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)
+		return fmt.Errorf("(reason:trace_id_zero): %s", s)
```

Comment:
These error messages should be user friendly and good to be used anywhere. They should also be human readable. The way they are written now they seem like they are expected to be appended to another string, which they are, but they should work on their own. Please make all of them human friendly:
```suggestion
		return fmt.Errorf("TraceID is zero: %s", s)
```

Reply:
Wouldn't it be better to maintain consistency with the metric tags, log warning output, and `datadog-agent status` output? The main reason anyone would use these logs is if they saw either the dropped/malformed trace metrics, or saw the periodic summary log message, in which case isn't it more user friendly to maintain the exact same strings? I imagine most people would do a grep for `trace_id_zero` to find the right log message? 

As for the `(reason:trace_id_zero)` formatting, the goal is to maintain a consistent format forall of these trace normalization debug log messages:

*`Fixing malformed trace (reason:invalid_http_status_code) ...`
*`Dropping invalid trace (reason:trace_id_zero) ...`

That's why initially I figured it's simplest to do the logging right where the decision is being made. I moved it out though so the error now seems a bit out of place, but the log messages will at least be consistent.

Reply:
Why does logging need to be consistent with tags used in stats? That's the part I don't get. These logs are mostly for humans, stats are written in a more cryptic fashion due to their limitations. To me `fmt.Errorf("(reason:trace_id_zero): %s", s)` does not look like a nice error. The error should make sense on its own and shouldn't make assumptions about the context.

Reply:
> Why does logging need to be consistent with tags used in stats? That's the part I don't get. These logs are mostly for humans, stats are written in a more cryptic fashion due to their limitations.

That is precisely why it needs to match. Because this is what a *human* is going to experience:

1. sees this in the periodic summary logs: `WARN: malformed_traces(empty_span_name: 4, invalid_duration:3) Enable debug logging for more information` 
1. enables debug logging
1. Greps logs for `empty_span_name` or `invalid_duration`, finds more inforamtion
1. Goes to metrics to plot when this started using tags `reason:empty_span_name`, `reason:invalid_duration`
1. ...

It's going to make the human's life much easier if all of that is consistent. 

> To me `fmt.Errorf("(reason:trace_id_zero): %s", s)` does not look like a nice error. The error should make sense on its own and shouldn't make assumptions about the context.

Ok, I updated the error and logging to stand better on its own Alright, updated here https://github.com/DataDog/datadog-agent/pull/3738/commits/6ff5cf52cc4ce0b718cbdac1a3f4a094e80abcd0

However, I really think we should keep the reason names consistent here with the rest of the logging and metrics.

Reply:
Ok, let's make a mixture of both then, for example:
```go
return fmt.Errorf("TraceID is zero (reason:trace_id_zero)")
```
Would you be happy with that? The same for all the others.

Reply:
Do we really need that though? It's adding unnecessary duplication and making it more verbose for no real benefit:

* `Trace is Empty (reason:empty_trace)`
* `Trace has foreign span (reason:foreign_span)`
* `Trace has duplicate span ID (reason:duplicate_span_id)`
* `Trace has invalid service (reason:service_invalid)`
* ...